### ISSUE TITLE
Add "since=" argument to note_deprecation, and a 'sleep incentive'

### ISSUE THREAD

**Issue by Zac-HD** [HypothesisWorks/hypothesis#1697]
[The "sleep constructor"](https://twitter.com/johnregehr/status/920691341738123264) is a highly effective way to stop people using an interface without actually breaking compatibility: when the deprecated function is called, you emit a warning and then sleep for a few seconds before doing the thing.  I think we should do this, and automatically increase the sleep period over time.

I originally suggested this for v4 in #1696, but (a) why wait? and (b), why make that diff any larger than it needs to be?

- [ ] Add a `since=` argument to `note_deprecation`, which takes a date in `"YYYY-MM-DD"` format.
- [ ] The default value should be the date of the most recent new deprecation; at time of writing [3.82.0](https://hypothesis.readthedocs.io/en/latest/changes.html#v3-82-0).
  Optionally, add the real date of deprecation to various calls and backdate the default accordingly.  Don't waste too much time on this though as most will be deleted for Hypothesis 4.0 in a few months and we can easily make it a required argument then.
- [ ] Set up our CI release function (in `tooling/`) to replace `since="RELEASEDAY"` with the then-current date string so that we get it right automatically.  Ensure that `note_deprecation` does the same substitution at runtime to support testing unmerged PRs.
- [ ] In `note_deprecation`, calculate the days since deprecation and sleep length `min(max(5, days * .05), 30)` seconds.  This works OK with badly-set clocks, is steady for an initial three month window before growing, and is enough to strongly encourage migration without being totally crippling.
- [ ] Append `"\nSleeping for %s seconds to encourage migration"` to the message, warn, then sleep *after* emitting the message.
- [ ] Monkey-patch this out in `conftest.py` so our own tests of deprecated behavior aren't affected.

**Comment by DRMacIver** [comment]
I'm not actually wild about the sleep incentive idea. It's tempting but I think it will burn more goodwill than it will achieve useful results.

**Comment by Zac-HD** [comment]
Why is the funnest stuff always a bad idea? Ah well...

Just the CI bit to insert the appropriate date then, and an assertion that the since argument is a valid date or the releaseday string.

**Comment by DRMacIver** [comment]
> Why is the funnest stuff always a bad idea? Ah well...

[![image](https://user-images.githubusercontent.com/28623/50234580-eb18b000-03ad-11e9-9169-1961f49beb47.png)](http://www.girlgeniusonline.com/comic.php?date=20031017)

**Comment by kxepal** [comment]
I agree, that this sleep-constructor idea is funny, but not good. People could easily avoid sleeping release and remain with the old one, because hypothesis just works. 

Another issues will be caused for public CI services. Are we sure that such sleeping tests will fail with proper error when CI will abort them because of execution timeout? Hypothesis tests are tend to run longer. Such change may let them run even more longer.

I think it would be more than enough a friendly (stdlib) warning and follow well known semantic versioning (may be not stricly, but with the similar expectations) when all the deprecated things turns into pumpkit with the next major release.
